rtic-time/src/monotonic.rs

```diff
@@ -22,203 +25,29 @@ pub trait Monotonic {
-                use ::fugit::ExtU64Ceil;
-                Self::delay(u64::from(ms).millis_at_least()).await;
-            }
-        }
-    };
-}
-
-/// Creates impl blocks for `embedded_hal::delay::DelayUs`,
-/// based on `fugit::ExtU32Ceil`.
-#[macro_export]
-macro_rules! embedded_hal_delay_impl_fugit32 {
-    ($t:ty) => {
-        impl ::embedded_hal::delay::DelayNs for $t {
-            fn delay_ns(&mut self, ns: u32) {
-                use ::fugit::ExtU32Ceil;
-
-                let now = Self::now();
-                let mut done = now + ns.nanos_at_least();
-                if now != done {
-                    // Compensate for sub-tick uncertainty
-                    done += Self::TICK_PERIOD;
-                }
-
-                while Self::now() < done {}
-            }
-
-            fn delay_us(&mut self, us: u32) {
-                use ::fugit::ExtU32Ceil;
-
-                let now = Self::now();
-                let mut done = now + us.micros_at_least();
-                if now != done {
-                    // Compensate for sub-tick uncertainty
-                    done += Self::TICK_PERIOD;
-                }
-
-                while Self::now() < done {}
-            }
-
-            fn delay_ms(&mut self, ms: u32) {
-                use ::fugit::ExtU32Ceil;
-
-                let now = Self::now();
-                let mut done = now + ms.millis_at_least();
-                if now != done {
-                    // Compensate for sub-tick uncertainty
-                    done += Self::TICK_PERIOD;
-                }
-
-                while Self::now() < done {}
-            }
-        }
-    };
-}
+    /// Delay for some duration of time.
+    async fn delay(duration: Self::Duration);
 
-/// Creates impl blocks for `embedded_hal_async::delay::DelayUs`,
-/// based on `fugit::ExtU32Ceil`.
-#[macro_export]
-macro_rules! embedded_hal_async_delay_impl_fugit32 {
-    ($t:ty) => {
-        impl ::embedded_hal_async::delay::DelayNs for $t {
-            #[inline]
-            async fn delay_ns(&mut self, ns: u32) {
-                use ::fugit::ExtU32Ceil;
-                Self::delay(ns.nanos_at_least()).await;
-            }
+    /// Delay to some specific time instant.
+    async fn delay_until(instant: Self::Instant);
 
-            #[inline]
-            async fn delay_us(&mut self, us: u32) {
-                use ::fugit::ExtU32Ceil;
-                Self::delay(us.micros_at_least()).await;
-            }
+    /// Timeout at a specific time.
+    async fn timeout_at<F: core::future::Future>(
+        instant: Self::Instant,
+        future: F,
+    ) -> Result<F::Output, TimeoutError>;
 
-            #[inline]
-            async fn delay_ms(&mut self, ms: u32) {
-                use ::fugit::ExtU32Ceil;
-                Self::delay(ms.millis_at_least()).await;
-            }
-        }
-    };
+    /// Timeout after a specific duration.
+    async fn timeout_after<F: core::future::Future>(
```

Comment:
After playing a bit it feels like we should add an `interval` method here that does the heavy lifting of `timeout_at` and manages the instances. Seems a too common usecase to leave out. :) 

Thoughts?

Reply:
I agree :) although this would only be considered a minor change according to semver, so I might suggest we add it afterwards. This pr is already huge as-is. Maybe open an issue for it? Sounds like a good beginner task.

Reply:
What do you mean with "manages the instances"?

Reply:
I tried to keep this pr a pure API rework; your suggestion would introduce new functionality.

Reply:
Yeah it's fine to not add it :) The extension of functionality we can add later.